x/ibc/testing/chain.go

```diff
@@ -0,0 +1,306 @@
+		Vals:     valSet,
+		Signers:  signers,
+		Querier:  keeper.NewQuerier(*app.IBCKeeper),
+	}
+}
+
+// GetContext creates a simple context for testing purposes.
+func (chain *TestChain) GetContext() sdk.Context {
+	return chain.App.BaseApp.NewContext(false, abci.Header{ChainID: chain.Header.SignedHeader.Header.ChainID, Height: int64(chain.Header.GetHeight())})
+}
+
+// CreateClient will create a client on the chain using the provided counterparty
+// TestChain.
+func (chain *TestChain) CreateClient(counterparty *TestChain) error {
+	// store the current header so we can properly store its info on the counterparty chain and client
+	header := counterparty.Header
+
+	// commit and create a new block on the counterparty chain to get a fresh CommitID
+	counterparty.App.Commit()
+	commitID := counterparty.App.LastCommitID()
+
+	// update counterparty with new header so we can begin a new block
+	counterparty.Header = nextHeader(counterparty)
+	counterparty.App.BeginBlock(abci.RequestBeginBlock{Header: abci.Header{Height: counterparty.Header.Height, Time: counterparty.Header.Time}})
+
+	// set HistoricalInfo on counterparty chain after Commit
+	ctxCounterparty := counterparty.GetContext()
+	validator := stakingtypes.NewValidator(
+		sdk.ValAddress(counterparty.Vals.Validators[0].Address), counterparty.Vals.Validators[0].PubKey, stakingtypes.Description{},
+	)
+	validator.Status = sdk.Bonded
+	validator.Tokens = sdk.NewInt(1000000) // get one voting power
+	validators := []stakingtypes.Validator{validator}
+	histInfo := stakingtypes.HistoricalInfo{
+		Header: abci.Header{
+			Time:    header.Time,
+			AppHash: commitID.Hash,
+		},
+		Valset: validators,
+	}
+	counterparty.App.StakingKeeper.SetHistoricalInfo(ctxCounterparty, header.Height, histInfo)
+
+	// also set staking params
+	stakingParams := stakingtypes.DefaultParams()
+	stakingParams.HistoricalEntries = 10
+	counterparty.App.StakingKeeper.SetParams(ctxCounterparty, stakingParams)
+
+	// create target ctx
+	ctxTarget := chain.GetContext()
+
+	// create client
+	clientState, err := ibctmtypes.Initialize(counterparty.ClientID, lite.DefaultTrustLevel, TrustingPeriod, UnbondingPeriod, MaxClockDrift, header)
+	if err != nil {
+		return err
+	}
+	_, err = chain.App.IBCKeeper.ClientKeeper.CreateClient(ctxTarget, clientState, header.ConsensusState())
+	if err != nil {
+		return err
+	}
+	return nil
+
+	// TODO: use simapp SignCheckDeliver to update state
+	// _, _, err := simapp.SignCheckDeliver(
+	// 	suite.T(),
+	// 	suite.cdc,
+	// 	suite.app.BaseApp,
+	// 	ctx.BlockHeader(),
+	// 	[]sdk.Msg{clienttypes.NewMsgCreateClient(clientID, clientexported.ClientTypeTendermint, consState, accountAddress)},
+	// 	[]uint64{baseAccount.GetAccountNumber()},
+	// 	[]uint64{baseAccount.GetSequence()},
+	// 	true, true, accountPrivKey,
+	// )
+}
+
+// UpdateClient will update a client on the chain using the provided counterparty
+// TestChain.
+func (chain *TestChain) UpdateClient(counterparty *TestChain) {
```

Comment:
So this cleans up the hacky work that me and Fede did with TestChain a while back, but ideally we don't do this at all. This function is mocking the functionality of `UpdateClient`, but it would be better to actually construct the update client msg from `counterparty` and deliver the msg to the target chain if possible.

I know that probably involves a pretty intense rewrite (hence why I didn't do it before), but if it's possible I think it's worth tackling this during a massive testing rewrite anyway. Would be a much more robust testing framework them

Reply:
I'll refactor to use message passing